# Apple Health Mobility Report

Our goal is to build a report with the Apple Health data that can be used by Physical Therapists to assess the progress of their patients. To do this, we selected some exercise metrics that we believe have an impact in maintaining or improving the mobility of people affected by Multiple Sclerosis (MS) or stroke.

1. Distance Cycling
2. Double support time
3. Exercise minutes
4. Six minute walk
5. Stair speed down
6. Stair speed up
7. Step length
8. Walking and running distance
9. Walking assymetry
10. Walking speed
11. Walking steadiness (last 4 weeks)

These metrics are a part of [Apple's Healthkit](https://developer.apple.com/documentation/healthkit) and the user can access them in individual charts. There is no native option on the iPhone and looking at the metrics in separate charts makes it harder to find correlations or justify discrepancies in the data.

This report aims to gather the above metrics in a way that can be analysed further and be more useful to understand the patient's progress. 

To do this we will use the full export of the Apple Health App and [convert it to a Pandas Dataframe using Milind Ganjoo's Apple Health Exporter code](https://github.com/mganjoo/apple-health-exporter).

## Getting the user data to display

1. Export Apple Health data from within the Health app.

![click on "user" icon and then on "export health data"](images/exporting.png)

In [74]:
%%capture

import os 
import sys 

!{sys.executable} -m pip install --upgrade pip
!{sys.executable} -m pip install pandas pyarrow
!{sys.executable} -m pip install matplotlib


Proceeding with default values

In [75]:
!{sys.executable} -m pip install apple-health-exporter

from apple_health_exporter import health_xml_to_feather
output_file='patient-data/data.feather'
health_xml_to_feather(zip_file='patient-data/export.zip',output_file=output_file)

import pandas as pd
data = pd.read_feather(output_file)

# Metrics and their respective data types

1. **Distance Cycling** : HKQuantityTypeIdentifierDistanceCycling
2. **Double support time** : HKQuantityTypeIdentifierWalkingDoubleSupportPercentage
3. **Exercise minutes** : HKQuantityTypeIdentifierAppleExerciseTime
4. **Six minute walk** : HKQuantityTypeIdentifierSixMinuteWalkTestDistance
5. **Stair speed down** : HKQuantityTypeIdentifierStairDescentSpeed
6. **Stair speed up** : HKQuantityTypeIdentifierStairAscentSpeed
7. **Step length** : HKQuantityTypeIdentifierWalkingStepLength
8. **Walking and running distance** : HKQuantityTypeIdentifierDistanceWalkingRunning
9. **Walking assymetry** : HKQuantityTypeIdentifierWalkingAsymmetryPercentage 
10. **Walking speed** : HKQuantityTypeIdentifierWalkingSpeed 
11. **Walking steadiness** (last 4 weeks) : HKQuantityTypeIdentifierAppleWalkingSteadiness


## Distance cycling

This value is calculated using the iPhone's GPS every time the user registers a Cycling Workout.


## Double support time 

> This is the percentage of time during a walk that both feet are on the ground. A lower value means you spend more of your walk with your weight on one foot instead of two, which can be a sign of better balance. During a typical walk this measure will fall between 20 and 40%.
>
> Walking requires strength and coordination. Changes in these can affect your balance and two-foot contact time.
>
> Double support time naturally varies with how fast vou walk and the terrain, but may increase with age.
>
> Double support time is recorded automatically on iPhone when you carry your phone near vour waist, such as in a trouser pocket, and walk steadily over flat ground.

**Source:** Apple Health App on iOS 15.4

## Exercise minutes

> Every full minute of movement equal to or
exceeding the intensity of a brisk walk for vou
counts towards our dailv Exercise minutes.

**Source:** Apple Health App on iOS 15.4

## Six minute walk

> This is an estimate of how far you can walk on
flat ground in six minutes based on a week of
vour motion and workout data. The fitter vou
are, the higher this number will be.
>
> This prediction is considered a measure of your
capacity for exercise. Longer distances are
associated with healthier cardiac, respiratory,
circulatory and neuromuscular function, and
may reflect your ability to perform everyday
activities. Improvements in fitness and mobility
can increase our six-minute walk distance.
>
> The standard six-minute walk test measures
the maximum number of metres you can walk
on an unobstructed, flat course doing laps of a
standard length. The test is often given to
those who are recovering from a cardiac or
respiratory condition. You can get a predicted
six-minute walk distance up to 500 metres
using data from a device, such as Apple Watch.

**Source:** Apple Health App on iOS 15.4

## Stair speed down

> This is a measurement of how fast vou walk
down the stairs.
>
> Descending stairs safely is an important
indicator of our overall mobility.
> Descending stairs requires strength and
coordination. Changes in these factors, or pain,
can affect your speed.
> Apple Watch records your vertical speed when
you go down a flight that is at least 3 metres
(10 feet).

**Source:** Apple Health App on iOS 15.4


## Stair speed up 

>This is a measurement of how fast vou walk up
the stairs.
>
> Climbing stairs quickly can be a sign of strong
cardiovascular and musculoskeletal health.
>
> Walking up stairs requires strength,
coordination and aerobic fitness. Changes in
these factors, or pain, can affect your speed.
Apple Watch records your vertical speed when
you go up a flight that is at least 3 metres (10
feet).

**Source:** Apple Health App on iOS 15.4

## Step lenght 

## Walking and running distance

## Walking assymetry

> In a healthy walking pattern, the timing of the steps you take with each foot are very similar. Walking asymmetry is the per cent of time your steps with one foot are faster or slower than the other foot. This means the lower the percentage of asymmetry, the healthier your walking pattern.
>
> Uneven walking patterns, such as limping, can
be a sign of disease, injury or other health
issues. An even or symmetrical walk is often an
important physical theraov goal when
recovering from injury.
> 
> Your iPhone can record walking asymmetry
automaticallv when vou carrv vour phone near
your waist - such as in a pocket - and walk
steadilv on flat ground.

**Source:** Apple Health App on iOS 15.4



## Walking speed



## Walking steadiness

> This is an estimate of your stability while
walking. iPhone calculates Walking Steadiness
using your walking speed, step length, double
support time and walking asymmetry data
that's stored in Health. This provides a sense of
the way you walk.
>
> Your steadiness is also related to your fall risk.
As steadiness goes down, your risk of falling
goes up. Walking Steadiness is not an
indication of how likelv vou are to fall in anv
given moment, but an overall sense of fall risk
in the next 12 months.
> 
> iPhone can automatically record your Walking
Steadiness in Health when you carry your
iPhone in a pocket or holder near our waist.
You can also receive a notification if vour
Walking Steadiness is low or very low. You can
manage these notifications in Health Checklist.

**Source:** Apple Health App on iOS 15.4




First we need to know what sort of data we are working with. The command will show what columns we have in the DataFrame and their data type (dtype).

In [76]:
data.dtypes

type                       object
sourceName                 object
unit                       object
startDate     datetime64[ns, UTC]
endDate       datetime64[ns, UTC]
value                     float64
dtype: object

The `type` column identifies the sort of data that we have available, as listed before in [Metrics and their respective data types](#Metrics-and-their-respective-data-types).

In [77]:
data.groupby("type").size()

type
HKDataTypeSleepDurationGoal                                    1
HKQuantityTypeIdentifierActiveEnergyBurned                704272
HKQuantityTypeIdentifierAppleExerciseTime                  21346
HKQuantityTypeIdentifierAppleStandTime                     20480
HKQuantityTypeIdentifierAppleWalkingSteadiness                25
HKQuantityTypeIdentifierBasalEnergyBurned                 214430
HKQuantityTypeIdentifierBloodAlcoholContent                   11
HKQuantityTypeIdentifierBodyFatPercentage                   2505
HKQuantityTypeIdentifierBodyMass                            2512
HKQuantityTypeIdentifierBodyMassIndex                       3358
HKQuantityTypeIdentifierBodyTemperature                        1
HKQuantityTypeIdentifierDietaryCaffeine                       89
HKQuantityTypeIdentifierDietaryCarbohydrates                 161
HKQuantityTypeIdentifierDietaryEnergyConsumed                163
HKQuantityTypeIdentifierDietaryFatTotal                      130
HKQuantityTypeIdenti

As a first exercise, we want to get the walking distance for the last 30 days and plot that chart to see if it matches with the user's data on the iPhone. We are calculating per week and using sunday as the start of the week to keep aligned with how the HealthKit shows its charts.

In [78]:
walking_distance = data.loc[data['type'] == 'HKQuantityTypeIdentifierDistanceWalkingRunning'][['startDate','endDate','sourceName','unit','value']]

walking_distance

startDate                   endDate        sourceName  \
374809 2021-01-22 11:54:01+00:00 2021-01-22 11:54:04+00:00       Chronograph   
374810 2021-01-22 12:01:26+00:00 2021-01-22 12:01:44+00:00       Chronograph   
374811 2021-01-22 12:03:03+00:00 2021-01-22 12:04:04+00:00       Chronograph   
374812 2021-01-22 12:48:19+00:00 2021-01-22 12:48:35+00:00       Chronograph   
374813 2021-01-22 12:48:47+00:00 2021-01-22 12:48:50+00:00       Chronograph   
...                          ...                       ...               ...   
519571 2022-03-21 21:12:28+00:00 2022-03-21 21:12:43+00:00       Chronograph   
519572 2022-03-21 21:13:14+00:00 2022-03-21 21:13:35+00:00       Chronograph   
519573 2022-03-21 21:17:24+00:00 2022-03-21 21:17:35+00:00       Chronograph   
519574 2022-03-21 21:18:46+00:00 2022-03-21 21:18:49+00:00       Chronograph   
519575 2022-03-21 21:20:41+00:00 2022-03-21 21:20:44+00:00  SonicScrewdriver   

       unit     value  
374809   km  0.001425  
374810   km  0.032048  
374811   km  0.004288  
374812   km  0.034540  
374813   km  0.004893  
...     ...       ...  
519571   km  0.005877  
519572   km  0.008547  
519573   km  0.013533  
519574   km  0.005698  
519575   km  0.004320  

[144767 rows x 5 columns]

In [79]:
walking_distance_day=walking_distance.groupby(pd.Grouper(key='endDate', freq='D')).sum()
# walking_distance_day.loc[walking_distance_day['endDate'] == '2022-03']
walking_distance_day

value
endDate                            
2015-10-13 00:00:00+00:00  1.096450
2015-10-14 00:00:00+00:00  0.000000
2015-10-15 00:00:00+00:00  0.000000
2015-10-16 00:00:00+00:00  0.000000
2015-10-17 00:00:00+00:00  0.000000
...                             ...
2022-03-17 00:00:00+00:00  2.436473
2022-03-18 00:00:00+00:00  4.001712
2022-03-19 00:00:00+00:00  2.260375
2022-03-20 00:00:00+00:00  2.723850
2022-03-21 00:00:00+00:00  2.078089

[2352 rows x 1 columns]

The values don't seem to match the informatin from the iOS Health App. For example 2022-03-21 00:00:00+00:00 shows a sum of 2.078 Km, the app shows a value of 1.6Km.

![](images/IMG_10B10CFE9641-1.jpeg)

In [80]:
import matplotlib.pyplot as plt

walking_distance.last('6M').plot()

TypeError: 'last' only supports a DatetimeIndex index

# Joining related data to assess mobility

https://www.apple.com/ca/healthcare/docs/site/Measuring_Walking_Quality_Through_iPhone_Mobility_Metrics.pdf

## Hypothesis 1, the total distance walked in a day impacts the walking assimetry

## Hypothesis 2, exercise minutes reduce walking assymetry the next day

## Hypothesis 3, distance cycled decreases walking assymetry 

## Hypothesis 4, too much exercise minutes increase walking assymetry


# LOST+FOUND

how to export to excel

```python
## ARTICLES
articles_json = pd.read_json('data/articles.json')
articles_json.link = articles_json.link.apply(html.unescape)
articles_json.summary = articles_json.summary.apply(html.unescape)
articles_json.to_excel('content/developers/articles_'+ datetime_string + '.xlsx')
```